### PR TITLE
Use runtime instread of dockerclient in container gc

### PR DESCRIPTION
**Pull request by feiskyer**
Refactor container garbage collection to use runtime interface instead.

This refactor is a pre-requirement for client/server runtime at #13768

### CODE REVIEW

**Comment by feiskyer** [comment]
CC @dchen1107 @yujuhong

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -57,6 +58,8 @@ type Runtime interface {
 	// specifies whether the runtime returns all containers including those already
 	// exited and dead containers (used for garbage collection).
 	GetPods(all bool) ([]*Pod, error)
+	// Garbage collection of dead containers
```
Comment (derekwaynecarr) [inline pkg/kubelet/container/runtime.go]: use proper go doc conventions.

```
// GarbageCollect removes dead containers using the specified criteria
```

File: pkg/kubelet/dockertools/container_gc.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
+/*
```
Comment (yujuhong) [inline pkg/kubelet/dockertools/container_gc.go]: Is there any functional change in this file?

File: pkg/kubelet/container/runtime.go [hunk 3]
```diff
@@ -57,6 +58,8 @@ type Runtime interface {
...
+	GarbageCollect(maxPerPodContainer, maxContainers int, minAge time.Duration) error
```
Comment (yujuhong) [inline pkg/kubelet/container/runtime.go]: Should we just pass the `ContainerGCPolicy` instead?

**Comment by yujuhong** [comment]
@vishh, this pushes garbage collection down to the container runtime. Is this in line with your plans to improve the disk manager? We may need a better GC policy that's generic for all the runtimes in the near future.

@yifan-gu, FYI, as this changes the runtime interface.

File: pkg/kubelet/container/runtime.go (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/container/runtime.go]: @derekwaynecarr thanks

File: pkg/kubelet/dockertools/container_gc.go (hunk 2 above)
Comment (feiskyer) [inline pkg/kubelet/dockertools/container_gc.go]: @yujuhong No. Just move it into docker itself.

File: pkg/kubelet/container/runtime.go (hunk 3 above)
Comment (feiskyer) [inline pkg/kubelet/container/runtime.go]: @yujuhong  Reasonable, interface updated.

**Comment by vishh** [comment]
@yujuhong: I haven't thought through disk management yet. We might end up exposing additional runtime APIs. Image management is specific to the runtime and so this PR makes sense overall. 
I have not reviewed the code.

**Comment by yujuhong** [comment]
@feiskyer, just a heads-up this may conflict with #14686, which hopefully will land soon.

File: pkg/kubelet/rkt/rkt.go [hunk 4]
```diff
@@ -1082,7 +1082,7 @@ func (r *runtime) GetContainerLogs(pod *api.Pod, containerID string, logOptions
 }
 
 // GarbageCollect collects the pods/containers. TODO(yifan): Enforce the gc policy.
-func (r *runtime) GarbageCollect() error {
+func (r *runtime) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy) error {
```
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: So if #14686 gets in first, then the change we need to make is to derive the `--grace-period` and `--expire-prepared` from the `gcPolicy.MinAge`. AFAICS we can just set both the the `gcPolicy.MinAge`

Comment (feiskyer) [inline pkg/kubelet/rkt/rkt.go]: @yifan-gu OK

**Comment by feiskyer** [comment]
@yujuhong OK, I'll rebase when #14686 gets merged

**Comment by yujuhong** [comment]
@k8s-bot ok to test

**Comment by yujuhong** [comment]
PR looks good to me overall. Will review again after rebase and apply the lgtm label. Thanks for the PR!

**Comment by feiskyer** [comment]
@yujuhong Shippable is so slow now

**Comment by yujuhong** [comment]
> @yujuhong Shippable is so slow now

Mergebot is also slow or not working :\ Sorry for the wait.

**Comment by feiskyer** [comment]
@yujuhong @yifan-gu rebased since #14686 has been merged

**Comment by yifan-gu** [comment]
LGTM

**Comment by yujuhong** [comment]
Thanks for rebasing.

**Pull request merged**